semgrep/semgrep/rule.py

```diff
@@ -68,6 +68,15 @@ def __init__(self, raw: YamlTree[YamlMap]) -> None:
         if any([lang in REGEX_ONLY_LANGUAGE_KEYS for lang in self._languages]):
             self._validate_none_language_rule()
 
+    def __eq__(self, other: object) -> bool:
+        if not isinstance(other, type(self)):
+            return False
+
+        return self._raw == other._raw
+
+    def __hash__(self) -> int:
+        return hash(self.id)
```

Comment:
I think this may be problematic given the above definition of `__eq__`. Since `id` is not guaranteed to be unique this means we could have rules that produce the same hashable value, but are not `==`. If we want something really quick and easy, although not particularly fast, we could do something like `hash(json.dumps(self._raw))` or something like that.

Reply:
@mschwager That's OK -- hash equivalence isn't treated as the same as equality.

For example:

```python3
class Foo:
    def __init__(self, a, b):
        self.a = a
        self.b = b

    def __eq__(self, other):
        if not isinstance(other, type(self)):
            return False
        return (self.a == other.a and self.b == other.b)

    def __hash__(self):
        return hash(self.a)

x = Foo("abc", "def")
y = Foo("abc", "xyz")
z = Foo(888, "def")

print(hash(x))
print(hash(y))
print(hash(z))
print(x == y)
print(x == z)
print(y == z)
```
returns
```
-4144991442677912395
-4144991442677912395
888
False
False
False
```

Equivalence never checks __hash__.  Membership does (a la `x in y`), but only as a performance improvement.  Hash collisions are, afaik, only a performance hit, never a logical mistake.  (The performance hit of doing the deep comparison between the raw dicts on id collision is less than the hashing and serializing the json on every comparison).

I believe the test failure you're seeing is actually the corrected behavior: the snapshot version appears to be producing two hits from identical rules in two files.

```
>>> with open("tests/e2e/snapshots/test_check/test_multiple_configs_file/results.json") as x:
...     y = x.read()
...
>>> import json
>>> foo = json.loads(y)
>>> foo["results"][2] == foo["results"][0]
True
```

I will push a new patch correcting the snapshot.

Reply:
Sorry @mschwager, I misread your concern.  dicts and sets also check equality, not just hashing.  Extending the above Foo class example:

```
bar = set()
bar.add(x)
bar.add(y)
bar.add(z)
print(len(bar))
bar.add(x)
bar.add(y)
print(len(bar))
```
returns
```
3
3
```.

Reply:
Ahh, TIL that hashable types rely on `__eq__` too! [`__hash__`](https://docs.python.org/3/reference/datamodel.html#object.__hash__) was a lengthy, but helpful read. Thanks for the clarifications! This LGTM :+1: